server/index.go

```diff
@@ -0,0 +1,258 @@
+}
+
+func (idx *Index) store(t pack.BlobType, id, pack backend.ID, offset, length uint, old bool) {
+	idx.pack[id.String()] = indexEntry{
+		tpe:    t,
+		packID: pack,
+		offset: offset,
+		length: length,
+		old:    old,
+	}
+}
+
+// Store remembers the id and pack in the index.
+func (idx *Index) Store(t pack.BlobType, id, pack backend.ID, offset, length uint) {
+	idx.m.Lock()
+	defer idx.m.Unlock()
+
+	debug.Log("Index.Store", "pack %v contains id %v (%v), offset %v, length %v",
+		pack.Str(), id.Str(), t, offset, length)
+
+	idx.store(t, id, pack, offset, length, false)
+}
+
+// Remove removes the pack ID from the index.
+func (idx *Index) Remove(packID backend.ID) {
+	idx.m.Lock()
+	defer idx.m.Unlock()
+
+	debug.Log("Index.Remove", "id %v removed", packID.Str())
+
+	s := packID.String()
+	if _, ok := idx.pack[s]; ok {
+		delete(idx.pack, s)
+	}
+}
+
+// Lookup returns the pack for the id.
+func (idx *Index) Lookup(id backend.ID) (packID backend.ID, tpe pack.BlobType, offset, length uint, err error) {
+	idx.m.Lock()
+	defer idx.m.Unlock()
+
+	if p, ok := idx.pack[id.String()]; ok {
+		debug.Log("Index.Lookup", "id %v found in pack %v at %d, length %d",
+			id.Str(), p.packID.Str(), p.offset, p.length)
+		return p.packID, p.tpe, p.offset, p.length, nil
+	}
+
+	debug.Log("Index.Lookup", "id %v not found", id.Str())
+	return nil, pack.Data, 0, 0, errors.New("id not found")
+}
+
+// Has returns true iff the id is listed in the index.
+func (idx *Index) Has(id backend.ID) bool {
+	_, _, _, _, err := idx.Lookup(id)
+	if err == nil {
+		return true
+	}
+
+	return false
+}
+
+// Merge loads all items from other into idx.
+func (idx *Index) Merge(other *Index) {
+	debug.Log("Index.Merge", "Merge index with %p", other)
+	idx.m.Lock()
+	defer idx.m.Unlock()
+
+	for k, v := range other.pack {
+		if _, ok := idx.pack[k]; ok {
+			debug.Log("Index.Merge", "index already has key %v, updating", k[:8])
+		}
+
+		idx.pack[k] = v
+	}
+	debug.Log("Index.Merge", "done merging index")
+}
+
+// Each returns a channel that yields all blobs known to the index. If done is
+// closed, the background goroutine terminates. This blocks any modification of
+// the index.
+func (idx *Index) Each(done chan struct{}) <-chan pack.Blob {
+	idx.m.Lock()
+
+	ch := make(chan pack.Blob)
+
+	go func() {
+		defer idx.m.Unlock()
+		defer func() {
+			close(ch)
+		}()
+
+		for ids, blob := range idx.pack {
+			id, err := backend.ParseID(ids)
+			if err != nil {
+				// ignore invalid IDs
+				continue
+			}
+
+			select {
+			case <-done:
+				return
+			case ch <- pack.Blob{
+				ID:     id,
+				Offset: blob.offset,
+				Type:   blob.tpe,
+				Length: uint32(blob.length),
+			}:
+			}
+		}
+	}()
+
+	return ch
+}
+
+// Count returns the number of blobs of type t in the index.
+func (idx *Index) Count(t pack.BlobType) (n uint) {
+	debug.Log("Index.Count", "counting blobs of type %v", t)
+	idx.m.Lock()
+	defer idx.m.Unlock()
+
+	for id, blob := range idx.pack {
+		if blob.tpe == t {
```

Comment:
Hm this doesn't seem super efficient, especially if being called multiple times with different arguments `t`. Maybe it's worth it to run this only once and then cache the counts for all types so that those `Count()` queries can be instant? Or is this usually only a small number of blobs anyway?